file_store/src/price_tracker.rs

```diff
@@ -0,0 +1,83 @@
+use crate::{Error, FileStore, FileType, Result, Settings};
+use chrono::{Duration, Utc};
+use futures::stream::{StreamExt, TryStreamExt};
+use helium_proto::{services::price_oracle::PriceOracleReportV1, BlockchainTokenTypeV1, Message};
+use std::{collections::HashMap, sync::Arc};
+
+use tokio::sync::RwLock;
+type Prices = Arc<RwLock<HashMap<BlockchainTokenTypeV1, u64>>>;
+
+pub struct PriceTracker {
+    prices: Prices,
+}
+
+impl PriceTracker {
+    pub async fn start(
+        settings: &Settings,
+        shutdown: triggered::Listener,
+    ) -> Result<(Self, impl std::future::Future<Output = Result>)> {
+        let prices = Arc::new(RwLock::new(HashMap::new()));
+
+        let prices_clone = prices.clone();
+        let settings_clone = settings.clone();
+        let handle = tokio::spawn(async move { run(settings_clone, prices_clone, shutdown).await });
+
+        Ok((PriceTracker { prices }, async move {
+            match handle.await {
+                Ok(Ok(())) => Ok(()),
+                Ok(Err(err)) => Err(err),
+                Err(err) => Err(Error::from(err)),
```

Comment:
for `Error::from` not return an `Err(value)` such that you still have to wrap it in an error result value?

Reply:
`Error::from` does not return a `Result` just the actual error, so it still needs to be made into a `Result` or do I not understand the question?